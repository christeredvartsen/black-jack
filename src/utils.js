const gameState = {
  BOTH_WINS: 'both win',
  BOTH_LOSE: 'both lose',
  P1_WINS: 'player 1 wins',
  P2_WINS: 'player 2 wins',
  P1_DRAWS: 'player 1 to draw',
  P2_DRAWS: 'player 2 to draw',
};

const getCardValue = ({ value }) => {
  let result;

  switch (value.toLowerCase()) {
    case 'j': // continue
    case 'q': // continue
    case 'k': result = 10; break;
    case 'a': result = 11; break;
    default: result = parseInt(value, 10);
  };

  return result;
};

const getCardsScore = (cards) => cards.reduce((score, card) => score + getCardValue(card), 0);

const getGameState = (p1, p2) => {
  if (p1 === 21 && p2 === 21) {
    return gameState.BOTH_WINS;
  } else if (p1 > 21 && p2 > 21) {
    return gameState.BOTH_LOSE;
  } else if (p2 > 21 || p1 === 21) {
    return gameState.P1_WINS;
  } else if (p1 > 21 || (p1 >= 17 && p2 > p1)) {
    return gameState.P2_WINS;
  } else if (p1 < 17) {
    return gameState.P1_DRAWS;
  } else {
    return gameState.P2_DRAWS;
  }
};

export {
  getCardValue,
  getCardsScore,
  gameState,
  getGameState,
};