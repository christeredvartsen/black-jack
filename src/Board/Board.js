import React from 'react';
import PropTypes from 'prop-types';
import Deck from '../Deck/Deck';
import GameResult from '../GameResult/GameResult';
import Player from '../Player/Player';
import RestartGame from '../RestartGame/RestartGame';
import { gameState as s } from '../utils';
import types from '../propTypes';
import './Board.scss';

const Board = ({ drawCard, gameState, reset, deck, player1, player2 }) => (
  <div>
    <Deck cards={deck} />
    <GameResult result={gameState} p1={player1.name} p2={player2.name} />
    <div className="players">
      <Player {...player1} drawCard={() => drawCard('player1')} draw={gameState === s.P1_DRAWS} />
      <Player {...player2} drawCard={() => drawCard('player2')} draw={gameState === s.P2_DRAWS} />
    </div>
    <RestartGame reset={reset} />
  </div>
);
Board.propTypes = {
  deck: types.cards.isRequired,
  gameState: PropTypes.string.isRequired,
  drawCard: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired,
  player1: types.player.isRequired,
  player2: types.player.isRequired
};

export default Board;