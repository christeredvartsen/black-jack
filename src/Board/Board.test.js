import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import each from 'jest-each';
import Board from './Board';
import { gameState as s } from '../utils';

describe('<Board />', () => {
  const player1 = {
    name: 'name1',
    score: 0,
    cards: [],
  };
  const player2 = {
    name: 'name2',
    score: 0,
    cards: [],
  };

  each([
    [s.P1_DRAWS, 'player1'],
    [s.P2_DRAWS, 'player2'],
  ]).test('players can draw cards', (state, callbackArgument) => {
    const draw = jest.fn();
    const component = mount(
      <Board
        deck={[]}
        player1={player1}
        player2={player2}
        drawCard={draw}
        reset={jest.fn()}
        gameState={state}
      />
    );
    component.find('Player button').simulate('click');

    expect(draw).toHaveBeenCalledTimes(1);
    expect(draw).toHaveBeenCalledWith(callbackArgument);

  });

  test('matches snapshot', () => {
    const tree = shallow(
      <Board
        deck={[]}
        player1={player1}
        player2={player2}
        drawCard={jest.fn()}
        reset={jest.fn()}
        gameState={s.P1_WINS}
      />
    )
    expect(toJson(tree)).toMatchSnapshot();
  });
});