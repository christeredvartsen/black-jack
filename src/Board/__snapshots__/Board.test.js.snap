// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> matches snapshot 1`] = `
<div>
  <Deck
    cards={Array []}
  />
  <GameResult
    p1="name1"
    p2="name2"
    result="player 1 wins"
  />
  <div
    className="players"
  >
    <Player
      cards={Array []}
      draw={false}
      drawCard={[Function]}
      name="name1"
      score={0}
    />
    <Player
      cards={Array []}
      draw={false}
      drawCard={[Function]}
      name="name2"
      score={0}
    />
  </div>
  <RestartGame
    reset={[MockFunction]}
  />
</div>
`;
